python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -208,14 +216,27 @@ def threadpool_computations(
             )
         yield from formatted_batch_iter
 
+    def threadpool_computations_finalize_fn(batch_iter):
+        if finalize_fn is None:
+            yield from batch_iter
+        yield from finalize_batches(batch_iter, finalize_fn=finalize_fn, stats=stats)
+
     if num_threadpool_workers > 0:
-        return make_async_gen(
+        collated_iter = make_async_gen(
             base_iterator=batch_iter,
-            fn=threadpool_computations,
+            fn=threadpool_computations_format_collate,
             num_workers=num_threadpool_workers,
         )
+        finalized_iter = make_async_gen(
+            base_iterator=collated_iter,
+            fn=threadpool_computations_finalize_fn,
+            num_workers=1,
```

Comment:
Actually I don't think we need this case at all. We should just be able to call `threadpool_computations_finalize_fn(collated_iter)` directly in all cases since the whole thing is being run in a separate thread anyways.

Otherwise we would be prefetching 1 extra batch beyond what's specified in `prefetch_batches`